File: 2024/6/part2.py
from copy import deepcopy
from tqdm import tqdm

raw = open('input.dat').readlines()
raw = [line.strip() for line in raw]

# Create a 2D Grid from the string input
grid = []
for line in raw:
    row = [char for char in line]
    grid.append(row)

n, m = len(grid), len(grid[0])
dir = {
    '^': (-1, 0),
    'v': (1, 0),
    '>': (0, 1),
    '<': (0, -1)
}
symb = {
    '^': '>',
    '>': 'v',
    'v': '<',
    '<': '^'
}

for i in range(n):
    for j in range(m):
        if grid[i][j] == '^':
            x, y = i, j
            break

def simulate(grid, x, y):
    visited = set()
    path = set()

    while True:
        cur_dir = grid[x][y]
        
        if (x, y, cur_dir) in path:
            return True
        path.add((x, y, cur_dir))
        visited.add((x, y))

        dx, dy = dir[cur_dir]
        x, y = x + dx, y + dy

        if x not in range(n) or y not in range(m):
            return False

        if grid[x][y] == '#':
            x, y = x - dx, y - dy
            grid[x][y] = symb[cur_dir]
        else:
            grid[x][y] = cur_dir

valid_positions = 0
for i in tqdm(range(n)):
    for j in range(m):
        if grid[i][j] == '.' and (i, j) != (x, y):
            temp_grid = deepcopy(grid)
            temp_grid[i][j] = '#'

            if simulate(temp_grid, x, y):
                valid_positions += 1

print(valid_positions)

File: 2024/go/6/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type GuardPosition rune

type Situation struct {
	guard GuardPosition
	x, y  int
}

const (
	GUA_UP    GuardPosition = '^'
	GUA_DOWN  GuardPosition = 'v'
	GUA_LEFT  GuardPosition = '<'
	GUA_RIGHT GuardPosition = '>'

	OBS = '#'
	EMP = '.'
)

func main() {
	file, err := os.Open("input.dat")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}

	scanner := bufio.NewScanner(file)

	// Load the grid
	grid := make([][]rune, 0)
	for scanner.Scan() {
		line := scanner.Text()

		grid = append(grid, []rune(line))
	}
	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	// Locate the guard's initial position
	guard := GUA_UP
	var guardX, guardY int
	for y, row := range grid {
		for x, cell := range row {
			if cell == rune(guard) {
				guardX, guardY = x, y
			}
		}
	}

	cnt := 0
	for y, row := range grid {
		for x, cell := range row {
			if cell == EMP {
				grid[y][x] = OBS
				if checkLoop(guard, guardX, guardY, grid) {
					cnt += 1
				}
				grid[y][x] = EMP
			}
		}
	}

	fmt.Println(cnt)
}

func checkLoop(guard GuardPosition, guardX, guardY int, grid [][]rune) bool {
	var visited map[Situation]bool = make(map[Situation]bool)
	for inbounds(guardX, guardY, len(grid[0]), len(grid)) {
		if visited[Situation{guard, guardX, guardY}] {
			return true
		}
		visited[Situation{guard, guardX, guardY}] = true

		newX, newY := moveGuard(guard, guardX, guardY)
		if inbounds(newX, newY, len(grid[0]), len(grid)) && grid[newY][newX] == OBS {
			guard = rotateGuard(guard)
			continue
		}
		guardX, guardY = newX, newY
	}

	return false
}

func rotateGuard(guard GuardPosition) GuardPosition {
	switch guard {
	case GUA_UP:
		return GUA_RIGHT
	case GUA_RIGHT:
		return GUA_DOWN
	case GUA_DOWN:
		return GUA_LEFT
	case GUA_LEFT:
		return GUA_UP
	}
	panic(fmt.Sprintf("Invalid guard position: %c", guard))
}

func moveGuard(guard GuardPosition, x, y int) (int, int) {
	switch guard {
	case GUA_UP:
		return x, y - 1
	case GUA_DOWN:
		return x, y + 1
	case GUA_LEFT:
		return x - 1, y
	case GUA_RIGHT:
		return x + 1, y
	}
	panic(fmt.Sprintf("Invalid guard position: %c", guard))
}

func inbounds(x, y, maxX, maxY int) bool {
	return x >= 0 && x < maxX && y >= 0 && y < maxY
}
